lineage/tests/test_BaumWelch.py

```diff
@@ -0,0 +1,57 @@
+""" Unit test file. """
+import unittest
+import numpy as np
+from ..StateDistribution import StateDistribution
+from ..UpwardRecursion import get_leaf_Normalizing_Factors, calculate_log_likelihood
+from ..BaumWelch import fit
+from ..LineageTree import LineageTree
+from ..tHMM import tHMM
+
+
+class TestBW(unittest.TestCase):
+    """ Unit tests for Baum-Welch methods. """
+        
+    def test_step(self):
+        """ This tests that one step of Baum-Welch increases the likelihood of the fit. """
+        
+        # pi: the initial probability vector
+        pi = np.array([0.6, 0.4], dtype="float")
+
+        # T: transition probability matrix
+        T = np.array([[0.85, 0.15],
+                      [0.15, 0.85]], dtype="float")
+
+        # State 0 parameters "Resistant"
+        state0 = 0
+        bern_p0 = 0.95
+        gamma_a0 = 5.0
+        gamma_scale0 = 1.0
+
+        # State 1 parameters "Susciptible"
+        state1 = 1
+        bern_p1 = 0.85
+        gamma_a1 = 10.0
+        gamma_scale1 = 2.0
+
+        state_obj0 = StateDistribution(state0, bern_p0, gamma_a0, gamma_scale0)
+        state_obj1 = StateDistribution(state1, bern_p1, gamma_a1, gamma_scale1)
+
+        E = [state_obj0, state_obj1]
+        
+        num = 10000
+
+        # Using an unpruned lineage to avoid unforseen issues
+        X = LineageTree(pi, T, E, num, prune_boolean=False)
+
+        tHMMobj = tHMM([X], numStates=2)  # build the tHMM class with X
+
+        LLbefore = calculate_log_likelihood(tHMMobj, get_leaf_Normalizing_Factors(tHMMobj))
```

Comment:
Both LL that I calculate are coming out to be -inf, even before running fitting.

Reply:
I remember that it was set to be -inf at first, so it improves over time, right @shak360?

Reply:
Yes, the initial condition is just -inf because that makes sure that it will always pass the first check that that the next step will always have a higher likelihood than the 0th step.

Reply:
But if you comment this out, the LL after running fitting for two steps is still -inf.

Reply:
Just fixed these. Getting the `NF` for the whole tree requires running three functions (`get_leaf_NFs`, `get_leaf_betas`, and `get_nonleaf_NFs_and_betas`) not just `get_leaf_NFs`.

Reply:
Likelihoods are operating as expected (only ever increase through each iteration). Now the error is back in the aforementioned ` assert NF_array[leaf_cell_idx] > 0.0`. Will keep looking, but issue seems to be with `MSD`.